adapters/madvertise/madvertise.go

```diff
@@ -0,0 +1,171 @@
+package madvertise
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"text/template"
+
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	endpointTemplate template.Template
+}
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	template, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
+
+	bidder := &adapter{
+		endpointTemplate: *template,
+	}
+
+	return bidder, nil
+}
+
+func getHeaders(request *openrtb2.BidRequest) http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("X-Openrtb-Version", "2.5")
+
+	if request.Device != nil {
+		if len(request.Device.UA) > 0 {
+			headers.Add("User-Agent", request.Device.UA)
+		}
+
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+		}
+	}
+
+	return headers
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	zoneID := ""
+	for _, imp := range request.Imp {
+		madvertiseExt, err := getImpressionExt(imp)
+		if err != nil {
+			fmt.Println(err)
+			return nil, []error{err}
+		}
+		if madvertiseExt.ZoneID != "" {
+			if len(madvertiseExt.ZoneID) < 7 {
+				return nil, []error{&errortypes.BadInput{
+					Message: fmt.Sprintf("The minLength of zone ID is 7; ImpID=%s", imp.ID),
+				}}
+			}
+			if zoneID == "" {
+				zoneID = madvertiseExt.ZoneID
+			} else if zoneID != madvertiseExt.ZoneID {
+				return nil, []error{&errortypes.BadInput{
+					Message: "There must be only one zone ID",
+				}}
+			}
+		} else {
+			return nil, []error{&errortypes.BadInput{
+				Message: fmt.Sprintf("ZoneId is empty; ImpID=%s", imp.ID),
+			}}
+		}
+	}
+
+	url, err := a.buildEndpointURL(zoneID)
+	if err != nil {
+		return nil, []error{err}
+	}
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestData := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     url,
+		Body:    requestJSON,
+		Headers: getHeaders(request),
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+}
+
+func getImpressionExt(imp openrtb2.Imp) (*openrtb_ext.ExtImpMadvertise, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("ext not provided; ImpID=%s", imp.ID),
```

Comment:
Can you please append the original error message for this error as well?

Reply:
Done.

[master 949c9595] append the original error message